icu4j/main/tests/core/src/com/ibm/icu/dev/test/format/MeasureUnitTest.java

```diff
@@ -3340,65 +3340,248 @@ public void TestInvalidIdentifiers() {
+                centimeterSecond1Sub, centimeterSecond1Sub.length);
+        String secondCubicMeterSub[] = {
+                "cubic-meter", "square-kilosecond"
+        };
+        verifyCompoundUnit(secondCubicMeter, "cubic-meter-square-kilosecond",
+                secondCubicMeterSub, secondCubicMeterSub.length);
+        String secondCentimeterSub[] = {
+                "centimeter", "square-kilosecond"
+        };
+        verifyCompoundUnit(secondCentimeter, "centimeter-square-kilosecond",
+                secondCentimeterSub, secondCentimeterSub.length);
+        String secondCentimeterPerKilometerSub[] = {
+                "centimeter", "square-kilosecond", "per-kilometer"
+        };
+        verifyCompoundUnit(secondCentimeterPerKilometer, "centimeter-square-kilosecond-per-kilometer",
+                secondCentimeterPerKilometerSub, secondCentimeterPerKilometerSub.length);
+
+        assertTrue("reordering equality", cubicMeterSecond1.equals(secondCubicMeter));
+        assertTrue("additional simple units inequality", !secondCubicMeter.equals(secondCentimeter));
+
+        // Don't allow get/set power or SI prefix on compound units
+        try {
+            meterSecond.getDimensionality();
+            fail("UnsupportedOperationException must be thrown");
+        } catch (UnsupportedOperationException e) {
+            // Expecting an exception to be thrown
+        }
+
+        try {
+            meterSecond.withDimensionality(3);
+            fail("UnsupportedOperationException must be thrown");
+        } catch (UnsupportedOperationException e) {
+            // Expecting an exception to be thrown
+        }
+
+        try {
+            meterSecond.getSIPrefix();
+            fail("UnsupportedOperationException must be thrown");
+        } catch (UnsupportedOperationException e) {
+            // Expecting an exception to be thrown
+        }
+
+        try {
+            meterSecond.withSIPrefix(MeasureUnit.SIPrefix.CENTI);
+            fail("UnsupportedOperationException must be thrown");
+        } catch (UnsupportedOperationException e) {
+            // Expecting an exception to be thrown
+        }
+
+        MeasureUnit footInch = MeasureUnit.forIdentifier("foot-and-inch");
+        MeasureUnit inchFoot = MeasureUnit.forIdentifier("inch-and-foot");
+
+        String footInchSub[] = {
+                "foot", "inch"
+        };
+        verifyMixedUnit(footInch, "foot-and-inch",
+                footInchSub, footInchSub.length);
+        String inchFootSub[] = {
+                "inch", "foot"
+        };
+        verifyMixedUnit(inchFoot, "inch-and-foot",
+                inchFootSub, inchFootSub.length);
+
+        assertTrue("order matters inequality", !footInch.equals(inchFoot));
+
+
+        MeasureUnit dimensionless  /* TODO(sffc): do you mean that we need default constructor for dimensionless measure unit? */ = MeasureUnit.forIdentifier("");
```

Comment:
Here and below: Default constructors are a C++ thing.  In Java, we can use `null` to represent the dimensionless unit (you can't do that in C++ since MeasureUnit is a value type).

In Java, why don't you throw an exception on `MeasureUnit.forIdentifier("")` so we don't have to worry about that case.

Reply:
Do you mean we do not allow a `dimensionless` unit in Java ???

Reply:
I think Shane means we can do this in Java:

MeasureUnit dimensionless = null;

We can't do this in C, since we aren't using pointers to instances allocated on the heap: i.e. they're "value types" not "reference types" (I hope I get the terminology right).

Reply:
> Do you mean we do not allow a `dimensionless` unit in Java ???

Yes.  We have precedent for this.  In Java, NoUnit.BASE is `null`, but in C++, it is `MeasureUnit()`.

https://github.com/unicode-org/icu/blob/8e04bf016d266dd2e7a99b15d0eb7f22cb58ce8c/icu4j/main/classes/core/src/com/ibm/icu/util/NoUnit.java#L24

https://github.com/unicode-org/icu/blob/8e04bf016d266dd2e7a99b15d0eb7f22cb58ce8c/icu4c/source/i18n/unicode/nounit.h#L47

If we can avoid a dimensionless unit object in Java, that's avoids a lot of edge cases, and we don't have to test for them.

Reply:
I made the dimensionless equals `null` however, I do not agree with this for many reason. Lets discuss it in the meeting today